crates/kitsune_p2p/kitsune_p2p/src/gossip/simple_bloom.rs

```diff
@@ -0,0 +1,548 @@
+    let k_num = u32::from_le_bytes(*arrayref::array_ref![bloom, 8, 4]);
+    let k1 = u64::from_le_bytes(*arrayref::array_ref![bloom, 12, 8]);
+    let k2 = u64::from_le_bytes(*arrayref::array_ref![bloom, 20, 8]);
+    let k3 = u64::from_le_bytes(*arrayref::array_ref![bloom, 28, 8]);
+    let k4 = u64::from_le_bytes(*arrayref::array_ref![bloom, 36, 8]);
+    let sip_keys = [(k1, k2), (k3, k4)];
+    bloomfilter::Bloom::from_existing(&bloom[44..], bitmap_bits, k_num, sip_keys)
+}
+
+mod step_2_local_sync_inner;
+use step_2_local_sync_inner::*;
+mod step_3_initiate_inner;
+use step_3_initiate_inner::*;
+mod step_4_com_loop_inner;
+use step_4_com_loop_inner::*;
+
+kitsune_p2p_types::write_codec_enum! {
+    /// SimpleBloom Gossip Wire Protocol Codec
+    codec GossipWire {
+        /// Initiate a round of gossip with a remote node
+        Initiate(0x10) {
+            filter.0: PoolBuf,
+        },
+
+        /// Accept an incoming round of gossip from a remote node
+        Accept(0x20) {
+            filter.0: PoolBuf,
+        },
+
+        /// Send a chunks of gossip meta op data,
+        /// if "finished" this will be the final chunk.
+        Chunk(0x30) {
+            finished.0: bool,
+            chunks.1: Vec<Arc<MetaOpData>>,
+        },
+    }
+}
+
+struct NodeInfo {
+    last_touch: std::time::Instant,
+    was_err: bool,
+}
+
+pub(crate) enum HowToConnect {
+    Con(Tx2ConHnd<wire::Wire>),
+    Url(TxUrl),
+}
+
+pub(crate) struct SimpleBloomModInner {
+    tuning_params: KitsuneP2pTuningParams,
+    space: Arc<KitsuneSpace>,
+    ep_hnd: Tx2EpHnd<wire::Wire>,
+    evt_sender: futures::channel::mpsc::Sender<event::KitsuneP2pEvent>,
+
+    local_agents: HashSet<Arc<KitsuneAgent>>,
+    local_bloom: BloomFilter,
+    local_data_map: DataMap,
+    local_key_set: KeySet,
+
+    remote_metrics: HashMap<Tx2Cert, NodeInfo>,
+
+    last_initiate_check: std::time::Instant,
+    initiate_tgt: Option<Tx2Cert>,
+
+    incoming: Vec<(Tx2ConHnd<wire::Wire>, GossipWire)>,
+
+    last_outgoing: std::time::Instant,
+    send_interval_ms: u64,
+    outgoing: Vec<(Tx2Cert, HowToConnect, GossipWire)>,
+}
+
+impl SimpleBloomModInner {
+    pub fn new(
+        tuning_params: KitsuneP2pTuningParams,
+        space: Arc<KitsuneSpace>,
+        ep_hnd: Tx2EpHnd<wire::Wire>,
+        evt_sender: futures::channel::mpsc::Sender<event::KitsuneP2pEvent>,
+    ) -> Self {
+        let send_interval_ms: u64 = (
+            // !*)&^$# cargo fmt...
+            16384.0    // max bytes in a gossip msg
+            * 8.0      // bits per byte
+            * 1000.0   // milliseconds
+            / 1024.0   // kbps
+            / 1024.0   // mbps
+            / tuning_params.gossip_output_target_mbps
+        ) as u64;
+
+        // pick an old instant for initialization
+        let old = std::time::Instant::now()
+            .checked_sub(std::time::Duration::from_secs(60 * 24))
+            .unwrap();
```

Comment:
24 minutes ago? Just checking if you didn't mean to make it 24 hours, since 24 minutes seems weirdly specific.

Reply:
lol, woops : ) fixed :+1: